### DISCUSSION TITLE
Parameter Binding Breaks `array_agg`

### DISCUSSION THREAD

**Discussion by luciengaitskell** [sqlalchemy/sqlalchemy#7719] [Usage Questions] [answered]
### Describe the bug

The input to `array_agg` is bound to a parameter, preventing Postgres from inferring the type. I'd like to force the columns to be explicitly listed in the SQL statement, so postgres is able to infer the type. Is it possible to disable parameter binding for a single subquery?

### To Reproduce

I have this select from part of a larger query:
```python
q = select(
    [
        MyTable.a(
            func.array_agg(
                [
                    MyTable.b,
                    MyTable.c,
                    MyTable.e,
                ],
            ).label('other')
        ),
    ]
).group_by(MyTable.a)
```

Compiling for Postgres without literal binding yields:
```sql
SELECT my_table.a AS a, array_agg(:param_1) AS other FROM my_table GROUP BY my_table.a
```

However this means Postgres cannot determine the type inside the `array_agg()`:
`Could not choose a best candidate function. You might need to add explicit type casts`

### Error

```
sqlalchemy.exc.DatabaseError: (pg8000.exceptions.DatabaseError) {'S': 'ERROR', 'V': 'ERROR', 'C': '42725', 'M': 'function array_agg(unknown) is not unique', 'H': 'Could not choose a best candid
ate function. You might need to add explicit type casts.', 'P': '1000', 'F': 'parse_func.c', 'L': '568', 'R': 'ParseFuncOrColumn'}
``` 

However, if I manually execute a command like this:
```sql
SELECT my_table.a AS a, array_agg((my_table.b, my_table.c, my_table.d)) AS other FROM my_table GROUP BY my_table.a
```
Postgres has no trouble properly executing.

### Versions

- OS: macOS 11.6
- Python: 3.9.7
- SQLAlchemy: 1.4.31
- Database: Postgres 13
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pg8000

**Comment by CaselIT** [comment]
Hi, 

It's not clear to me what the expected sql should be in that case.
What is t.a that's used as a function and as a value? What are t.b and t.e?

**Comment by luciengaitskell** [comment]
@CaselIT Apologies, I have added more information to the original post. The bind within the `array_agg` seems to prevent proper function.

**Comment by zzzeek** [comment, marked as answer]
your hand-SQL example is using a tuple literal.  array_agg() accepts only a single expression, not a comma-separated list, hence the error message.  

You can invoke using `sqlalchemy.tuple_()` or `sqlalchemy.dialects.postgresql.array()`, however to fetch it back it looks like things are getting weird with how the driver returns the data, but it seems to work with `sqlalchemy.dialects.postgresql.array`:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.dialects.postgresql import array
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    x = Column(Integer)
    y = Column(Integer)


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(x=1, y=2), A(x=3, y=4), A(x=5, y=6)])
s.commit()

stmt = select(A.id, func.array_agg(array([A.x, A.y]))).group_by(A.id)

print(s.execute(stmt).all())
```

**Reply by luciengaitskell** [reply]
Thank you for the explanation there.

I've been looking to combine multiple data types into the same array, and this highlighted it isn't supported by Postgres yet. It seems like Postgres has been silently converting the inner 'array' to a string instead due to the multiple types within. In that case, I should specify a `cast([...], text)` in SQLAlchemy to solve the ambiguous type problem.
